File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution { // dfs
    static int[] g_numbers;
    static int g_target;
    static int result;
    public int solution(int[] numbers, int target) {
        g_numbers = numbers;
        g_target = target;
        dfs(0,0);
        
        return result;
    }
    
    void dfs(int sum, int idx){
        if(idx== g_numbers.length){
            if(sum == g_target)
                result ++;
            return;
        }
        
        dfs(sum+(g_numbers[idx]*(-1)),idx+1);
        dfs(sum+(g_numbers[idx]),idx+1);
        
    }
}

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    return dfs(0, 0,numbers,target)

def dfs(curRes, idx,numbers,target):
    res = 0
    if idx == len(numbers) and curRes == target:
        return 1
    elif idx >=len(numbers):
        return 0
    else :
        res += dfs(curRes + numbers[idx],idx+1,numbers,target)
        res += dfs(curRes + (-1)*numbers[idx],idx+1,numbers,target)
        return res